### DISCUSSION TITLE
Test-driven development

### DISCUSSION THREAD

**Discussion by RoboMWM** [GriefPrevention/GriefPrevention#1129] [General]
Features should be written in gherkin-style, and tests should be written to validate scenarios.

Personally I have not been trained in TDD (and I guess Bukkit makes it difficult to do so?) so it's a bit difficult for me to write up a test first, and then write the implementing code...?

Anyways, any ideas/examples on this guideline would be ideal if anyone has and TDD/BDD experience

**Comment by Jikoo** [comment]
Would Hamcrest be sufficient? A lot of the Gherkin Java libs over double the boilerplate required.

**Comment by RoboMWM** [comment]
Well I wasn't looking necessarily at frameworks, just more of methodologies/practices to get started with TDD because so far I generally think of what I need to implement code-wise, and then turn that into the feature? I think? Idk if that's how I think or start out when making features but afaik TDD is where you write the test first, and then the code (which means you'd need to know what your classes and method names would be beforehand?) Or do you actually write said classes, just empty method bodies and such?

**Comment by Jikoo** [comment]
BDD in general seems like a good route for developing and reorganizing GP's features. Take one section of a feature, pull the code, describe taste, and slap into its own method with some form of coverage. Most BDD seems to go idea -> abstraction -> tests <-> implementation. The ideas are just more detailed/descriptive and structured user stories, so it plays well with the existing system.

AFAIK the issue with Gherkin TDD is that Gherkin doesn't exactly lend itself to programming. It's a great way to express the feature in plain language, but the tools for it often add a lot of extra dev time translating the language into actual tests.

Bukkit certainly does make things harder. I think the reason I've never gotten that into unit testing (properly) is that most of the things I'm interested in are relatively heavily tied to server internals, and Bukkit is very bad about decoupling that. Fortunately, Mockito and [MockBukkit](https://github.com/seeseemelk/MockBukkit) help fill in a lot of the gaps for API-only projects.
